adds photo carousel react component and first part of ts walkthrough … (#3513)
temporalio/documentation: docs/encyclopedia/event-history/go.mdx

keywords:
  - temporal application
  - event history
  - workflow design
  - business logic activities
  - temporal workers
  - temporal SDK tutorial
  - go sdk temporal guide
  - java sdk temporal guide
  - php sdk temporal guide
  - python sdk temporal guide
  - typescript sdk temporal guide
  - learning temporal workflows
  - developing with temporal
  - temporal workflow execution
  - temporal activity management
  - worker process execution
---

import PhotoCarousel from "@site/src/components/elements/PhotoCarousel";

With Temporal, your Workflows can seamlessly recover from crashes or failures without you having to write recovery logic code. This is made possible by the [Event History](https://docs.temporal.io/workflows#event-history), a complete and durable log of everything that has happened in the lifecycle of a Workflow Execution.

Temporal uses the Event History to not only store the current state of a Workflow Execution—it records every step taken along the way. "When your workflow code (business logic) needs to schedule an Activity Execution, it does not call the Activity function directly but generates a Command which is sent to server to do the scheduling."

> A [Command](httpsg://docs.temporal.io/workflows#command) is a unit of work that the Worker needs to execute and let the Temporal Service know what to do next. The Temporal Service will act on these Commands such as scheduling an Activity or starting a Timer. These Commands are then mapped to Events which are persisted in case of failure. For example, if the Worker crashes, the Temporal Service uses this information to replay the code and recreate the state of the Workflow Execution to what it was immediately before the crash. It then resumes progress from the point of failure as if the failure never occurred. This allows you, as a developer, to code as if this type of failure wasn't even a possibility.

Reviewer: - fix broken link
- added clarity how Workflow Tasks, Commands, and Worfklows relate to each other.

```suggestion
A [Command](https://docs.temporal.io/glossary#command) is a unit of work that the Worker needs to execute and let the Temporal Service know what to do next. Commands are generated in the Workflow Task when the Worker executes a Workflow. The Temporal Service will act on these Commands such as scheduling an Activity or starting a Timer. These Commands are then mapped to Events which are persisted in case of failure. For example, if the Worker crashes, the Temporal Service uses this information to replay the code and recreate the state of the Workflow Execution to what it was immediately before the crash. It then resumes progress from the point of failure as if the failure never occurred. This allows you, as a developer, to code as if this type of failure wasn't even a possibility.
```

    "The code walkthrough will now begin. In this Workflow Definition, setting the Start-to-Close Timeout and setting a variable are internal steps. That is, these steps don't require any interaction with the Temporal Service.",
    "The Worker then reaches a statement that does require interaction with the Temporal Service. In this case, it's a request to execute an Activity. This causes the Worker to issue a Command to the Temporal Service and provides the details needed. For example, the `ScheduleActivityTask` Command contains details such as the Task Queue name, the Activity Type, and the input parameter values.",
    "Afterwards, the Worker continues executing the Workflow code. The next line, highlighted here, evaluates a variable. Depending on the outcome, it may throw an exception, which would cause the Workflow to fail. However, this example assumes that this is a delivery for a nearby customer. The execution will continue.",

>     "The Worker now reaches the call to the Timer, which is another statement that involves interaction with the Temporal Service. This causes the Worker to issue another Command, one which requests that it start a Timer. The duration is one of the details specified in this Command. Further execution of this Workflow will now pause for 30 minutes until the Timer fires.",

Reviewer: ```suggestion
    "The Worker now reaches the call to start a Timer, which is another statement that involves interaction with the Temporal Service. This causes the Worker to issue another Command, one which requests the Temporal Service to start a Timer. The duration is one of the details specified in this Command. Further execution of this Workflow will now pause for 30 minutes until the Timer fires.",
```

    "The Temporal Service records another Event in response to the Worker accepting the Task: `ActivityTaskStarted`. This Event is colored pink to indicate that it's an indirect result of the Command. By the way, the `Start-to-Close` Timeout indicates the amount of time that the Activity has to complete. Also, another thing to note is that the `ActivityTaskStarted` Event is not written to the Event History until a Task closes, because the number of retry attempts is an attribute of the `ActivityTaskStarted` Event.",
    "The Worker executes the code within the Activity Definition, and when that method returns a result, the Worker sends a message to the Temporal Service, notifying it that the Task is complete. To reiterate, this is just a notification, not a Command, because it's not requesting that the Temporal Service do something that will allow the Workflow Execution to progress. In response to this notification, the Temporal Service records another Event: `ActivityTaskCompleted`.",
    "The next statement that results in a Command is the call to start a Timer. It issues a `StartTimer` Command.",

>     "The Temporal Service responds by starting a Timer for 30 minutes and logging a `TimerStarted` Event to the history. It is a direct result of the `StartTimer` Command.",

Reviewer: ```suggestion
    "The Temporal Service responds after starting a Timer for 30 minutes in the Service, logging a `TimerStarted` Event to the history. It is a direct result of the `StartTimer` Command.",
```

>     "After 30 minutes has elapsed, the Timer is fired, and the Temporal Service records the Event to the history: `TimerFired`. The Workflow Execution continues with the next statement, but this is an internal step, meaning that it does not interact with the Temporal Service.",

Reviewer: ```suggestion
    "After 30 minutes has elapsed, the Timer is fired on the Temporal Service, which it then records the Event `TimerFired` to the history. The Workflow Execution continues with the next statement, but this is an internal step, meaning that it does not interact with the Temporal Service.",
```

    "The next line is a request to start a Timer, so the Worker issues a Command to the Service - `StartTimer` - requesting that it starts a Timer. The Service starts the Timer, records an Event - `TimerStarted` - and then records another Event when the Timer fires - `TimerFired`.",
    "Now imagine that the Worker happens to crash once it reaches to next line, so another Worker takes over, using Replay to restore the current state before continuing execution of the lines that follow.",
    "The Worker then requests the Event History to replay it. Once the Worker has the Event History, the Worker determines the expected sequence of Commands needed to restore the current state. The Worker is expecting to encounter the `ScheduleActivityTask` and `StartTimer` Commands.",
    "As the Worker executes the code during Replay, it reaches the first call to execute an Activity and creates a `ScheduleActivityTask` Command.",

>     "This Command matches the one expected based on the Event History. It's not only the right type of Command, with the same details, but it also occurs at the right position in the sequence of expected Commands. Therefore, Replay proceeds.",

Reviewer: nit: replay doesn't _actually care_ about the details, it just result of the Command

```suggestion
    "This Command matches the one expected based on the Event History. It's not only the right type of Command, but it also occurs at the right position in the sequence of expected Commands. Therefore, Replay proceeds.",
```
